### PR TITLE
Implement some feature of directory table.

### PR DESCRIPTION
**Pull request by wenchaozhang-123**
…rectory table.

This commit, we support some directory features as following:
1. support copy to directory table, the grammar is such as: COPY BINARY DIRECTORY TABLE <directory_table_name> 'relative_path' to 'dest_path';
2. support copy from/to directory table program
3. allow create/drop index on directory table
4. support download mode in gpdirtable tool

fix #ISSUE_Number

---

### Change logs

_Describe your change clearly, including what problem is being solved or what feature is being added._

_If it has some breaking backward or forward compatibility, please clary._

### Why are the changes needed?

_Describe why the changes are necessary._

### Does this PR introduce any user-facing change?

_If yes, please clarify the previous behavior and the change this PR proposes._

### How was this patch tested?

_Please detail how the changes were tested, including manual tests and any relevant unit or integration tests._

### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [ ] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [ ] Sign the Contributor License Agreement as prompted for your first-time contribution(*One-time setup*).
- [ ] Learn the [coding contribution guide](https://cloudberrydb.org/contribute/code), including our code conventions, workflow and more.
- [ ] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [ ] Document changes.
- [ ] Add tests for the change
- [ ] Pass `make installcheck`
- [ ] Pass `make -C src/test installcheck-cbdb-parallel`
- [ ] Feel free to request `cloudberrydb/dev` team for review and approval when your PR is ready🥳

### CODE REVIEW

File: src/backend/commands/copy.c [hunk 1]
```diff
@@ -472,9 +472,18 @@ DoCopy(ParseState *pstate, const CopyStmt *stmt,
...
+				cstate = BeginCopyToDirectoryTable(pstate, stmt->filename, stmt->dirfilename,
+												   rel, stmt->is_program, options);
+			}
+
+			else
+			{
+				cstate = BeginCopyTo(pstate, rel, query, relid,
+									 stmt->filename, stmt->is_program,
+									 stmt->attlist, options);
+			}
```
Comment (avamingli) [inline src/backend/commands/copy.c:486]: Could rel be NULL? and if so, will it crash in else branch?

Comment (wenchaozhang-123) [reply src/backend/commands/copy.c:486]: No, BecinCopyTo can accept rel is NULL and it processed in logic.

File: gpMgmt/bin/gpdirtableload [hunk 2]
```diff
@@ -67,14 +67,20 @@ def parseargs():
...
+                                            'the table root directory, while in download '
+                                            'mode, means directory to download')
 
     parser.add_argument('--force-password-auth', default=False, action='store_true',
                         help='Force a password prompt')
 
     parser.add_argument('--host', default="localhost",
                         help='Host to connect to')
-    parser.add_argument('--input-file', help='Input files or directory')
+    parser.add_argument('--input-file', help='In upload mode, this means input files or '
```
Comment (my-ship-it) [inline gpMgmt/bin/gpdirtableload:81]: Wenchao, please add two options:
--tag
--overwrite_exist

File: src/test/regress/output/directory_table.source [hunk 3]
```diff
@@ -1006,23 +1021,37 @@ SELECT md5_equal('dir_table2', 'nation4');
...
+COPY BINARY dir_table1 TO '@abs_srcdir@/data/dir_table1';   -- fail
+ERROR:  COPY to directory table must specify the relative_path name.
+\COPY BINARY DIRECTORY TABLE dir_table1 'nation1' TO '@abs_srcdir@/data/nation1';   -- fail
+\copy: parse error at "TABLE"
+COPY BINARY DIRECTORY TABLE dir_table1 'nation1' TO '@abs_srcdir@/data/nation1';   -- OK
+\COPY BINARY DIRECTORY TABLE dir_table1 'nation2' TO stdin; -- fail
+\copy: parse error at "TABLE"
+COPY BINARY DIRECTORY TABLE dir_table1 'nation2' TO stdin; -- fail
+ERROR:  COPY to directory table must specify the target file name.
+\COPY BINARY DIRECTORY TABLE dir_table1 'nation2' TO PROGRAM 'gzip -c -1 > @abs_srcdir@/data/nation2.gz';   -- fail
```
Comment (my-ship-it) [inline src/test/regress/output/directory_table.source:1052]: enable  \COPY TO

**Review by my-ship-it — APPROVED** [review]

**Review by my-ship-it — APPROVED** [review]

**Pull request merged**